### ISSUE TITLE
Parameter shapeholder does not work with link copy.

### ISSUE THREAD

**Issue by delta2323** [pfnet/chainer#1615]
When we defer the initialization of parameters with the parameter shapeholder, they are created by the first forward propagation. If we copy the model before the first forward propagation, it does not create the parameter in the copied model. But as parameters are "marked" as initialized, there is no chance for copied model to create ones. This is problematic when we copy the link for evaluation, as we do in the examples.

Following code illustrates the problem

```
import chainer
from chainer import links as L
import numpy

x = numpy.random.uniform(size=(8, 16)).astype(numpy.float32)
x = chainer.Variable(x)

l = L.Linear(None, 4)
l_c = l.copy()  # (1)
l(x)  # (2)
l_c(x)
```

Traceback

```
Traceback (most recent call last):
  File "test.py", line 11, in <module>
    l_c(x)
  File "/Users/oonokenta/dev/chainer/chainer/links/connection/linear.py", line 86, in __call__
    return linear.linear(x, self.W, self.b)
AttributeError: 'Linear' object has no attribute 'W'
```

If we switch `(1)` and `(2)`, the error does not occur.

**Comment by zori** [comment]
Can confirm. deepcopy-ing the model will create it's own set of uninitialized parameters. See https://github.com/pfnet/chainer/pull/1576

**Comment by niboshi** [comment]
Closing as this has been fixed in #2072.
